src/Mir/Trans.hs

```diff
@@ -296,172 +296,140 @@ transBinOp bop op1 op2 = do
-              (M.Div, Just M.Signed) -> return $ MirExp (C.BVRepr n) (S.app $ E.BVSdiv n e1 e2)
-              (M.Rem, Just M.Unsigned) -> return $ MirExp (C.BVRepr n) (S.app $ E.BVUrem n e1 e2)
-              (M.Rem, Just M.Signed) -> return $ MirExp (C.BVRepr n) (S.app $ E.BVSrem n e1 e2)
-              (M.BitXor, _) -> return $ MirExp (C.BVRepr n) (S.app $ E.BVXor n e1 e2)
-              (M.BitAnd, _) -> return $ MirExp (C.BVRepr n) (S.app $ E.BVAnd n e1 e2)
-              (M.BitOr, _) -> return $ MirExp (C.BVRepr n) (S.app $ E.BVOr n e1 e2)
-              (M.Shl, _) ->
-                 let res = MirExp (C.BVRepr n) (S.app $ E.BVShl n e1 e2)
-                 -- TODO check unsigned vs signed???
-                 in extendUnsignedBV res na
-              (M.Shr, Just M.Unsigned) ->
-                 let res = MirExp (C.BVRepr n) (S.app $ E.BVLshr n e1 e2)
-                 in extendUnsignedBV res na
-              (M.Shr, Nothing) ->
-                 let res = MirExp (C.BVRepr n) (S.app $ E.BVLshr n e1 e2)
-                 in extendUnsignedBV res na
-              (M.Shr, Just M.Signed) ->
-                 let res = MirExp (C.BVRepr n) (S.app $ E.BVAshr n e1 e2) 
-                 in extendSignedBV res na
-              (M.Lt, Just M.Unsigned) -> return $ MirExp (C.BoolRepr) (S.app $ E.BVUlt n e1 e2)
-              (M.Lt, Just M.Signed)   -> return $ MirExp (C.BoolRepr) (S.app $ E.BVSlt n e1 e2)
-              (M.Le, Just M.Unsigned) -> return $ MirExp (C.BoolRepr) (S.app $ E.BVUle n e1 e2)
-              (M.Le, Just M.Signed)   -> return $ MirExp (C.BoolRepr) (S.app $ E.BVSle n e1 e2)
-
-              (M.Gt, Just M.Unsigned) -> return $ MirExp (C.BoolRepr) (S.app $ E.BVUlt n e2 e1)
-              (M.Gt, Just M.Signed)   -> return $ MirExp (C.BoolRepr) (S.app $ E.BVSlt n e2 e1)
-              (M.Ge, Just M.Unsigned) -> return $ MirExp (C.BoolRepr) (S.app $ E.BVUle n e2 e1)
-              (M.Ge, Just M.Signed)   -> return $ MirExp (C.BoolRepr) (S.app $ E.BVSle n e2 e1)
-
-              (M.Ne, _) -> return $ MirExp (C.BoolRepr) (S.app $ E.Not $ S.app $ E.BVEq n e1 e2)
-              (M.Beq, _) -> return $ MirExp (C.BoolRepr) (S.app $ E.BVEq n e1 e2)
+              (M.Add, _) -> do
+                let carry = case mat of
+                        Just M.Unsigned -> E.BVCarry
+                        Nothing -> E.BVCarry
+                        Just M.Signed -> E.BVSCarry
+                return (MirExp (C.BVRepr n) (S.app $ E.BVAdd n e1 e2), S.app $ carry n e1 e2)
+              (M.Sub, _) -> do
+                let borrow = case mat of
+                        Just M.Unsigned -> E.BVUlt
+                        Nothing -> E.BVUlt
+                        Just M.Signed -> E.BVSBorrow
+                return (MirExp (C.BVRepr n) (S.app $ E.BVSub n e1 e2), S.app $ borrow n e1 e2)
+              -- FIXME: implement overflow checks for Mul, Div, and Rem
```

Comment:
It would probably be good to create an issue for this before merging, so we don't forget.

Reply:
Opened #18 for the missing Mul/Div/Rem overflow checks